### PR TITLE
Blobs get TTL upload should be uploaded to Cloud.

### PR DESCRIPTION
**Pull request by zzmao**
Fix blobs with ttlUpdate upload to cloud bug.
Refactored test Classes in ReplicationTest for their use in other modules.
Fix metric name typo in HelixVcrClusterMetrics.

### CODE REVIEW

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java [hunk 1]
```diff
@@ -207,8 +207,14 @@ public void updateTtl(MessageWriteSet messageSetToUpdate) throws StoreException
     // Note: we skipped uploading the blob on PUT record if the TTL was below threshold.
     try {
       for (MessageInfo msgInfo : messageSetToUpdate.getMessageSetInfo()) {
-        BlobId blobId = (BlobId) msgInfo.getStoreKey();
-        cloudDestination.updateBlobExpiration(blobId, msgInfo.getExpirationTimeInMs());
+        // MessageInfo.expirationTimeInMs is not reliable if ttlUpdate is set. See {@code PersistentIndex#findKey()}
+        // and {@code PersistentIndex#markAsPermanent()}.
+        if (msgInfo.isTtlUpdated()) {
+          BlobId blobId = (BlobId) msgInfo.getStoreKey();
+          cloudDestination.updateBlobExpiration(blobId, Utils.Infinite_Time);
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java]: I was trying to keep the fact that UpdateTTL makes the blob permanent out of the cloud logic, but that may not be possible if expirationTime is not reliable.

File: ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java [hunk 2]
```diff
@@ -255,6 +279,156 @@ public void testExceptionalDest() throws Exception {
...
+   * Test PUT(with TTL) and TtlUpdate record replication.
+   * Replication may happen after PUT and after TtlUpdate, or after TtlUpdate only.
+   * PUT may already expired, expiration time < upload threshold or expiration time >= upload threshold.
+   * @throws Exception
+   */
+  @Test
+  public void testPutWithTtl() throws Exception {
+    // Set up remote host
+    MockClusterMap clusterMap = new MockClusterMap();
+    MockHost remoteHost = getLocalAndRemoteHosts(clusterMap).getSecond();
```
[the remark this answers is not in the document]
Comment (zzmao) [reply ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java]: `MockStore` is used by `remoteHost`(a regular data node).

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java (hunk 1 above)
Comment (zzmao) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java]: Yep. Current ttlUpdate expiration time is not always -1.

**Comment by zzmao** [comment]
Comment addressed. Let's try to merge it today.

File: ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lightningrob) [reply ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java]: Yeah, that makes sense.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java (hunk 1 above)
Comment (lightningrob) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java]: Maybe just add a comment that if we ever change updateTTL to be more flexible, that this code will need to be modified.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java [hunk 3]
```diff
@@ -174,10 +174,10 @@ private boolean shouldUpload(MessageInfo messageInfo) {
     if (messageInfo.isDeleted()) {
       return false;
     }
-
-    // expiration time above threshold
+    // expiration time above threshold. Expired blobs are blocked by ReplicaThread.
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java]: Update the comment to include "or is a ttl update"

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java [hunk 4]
```diff
@@ -207,8 +207,14 @@ public void updateTtl(MessageWriteSet messageSetToUpdate) throws StoreException
...
       for (MessageInfo msgInfo : messageSetToUpdate.getMessageSetInfo()) {
-        BlobId blobId = (BlobId) msgInfo.getStoreKey();
-        cloudDestination.updateBlobExpiration(blobId, msgInfo.getExpirationTimeInMs());
+        // MessageInfo.expirationTimeInMs is not reliable if ttlUpdate is set. See {@code PersistentIndex#findKey()}
+        // and {@code PersistentIndex#markAsPermanent()}.
+        if (msgInfo.isTtlUpdated()) {
+          BlobId blobId = (BlobId) msgInfo.getStoreKey();
+          cloudDestination.updateBlobExpiration(blobId, Utils.Infinite_Time);
+        } else {
+          logger.error("updateTtl() is called but msgInfo.isTtlUpdated is not set. msgInfo: {}", msgInfo);
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java]: Wonder if this should be IllegalStateException.  If that is too severe, then add a VcrMetric error counter to track number of this or similar errors that are encountered.

File: ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java [hunk 5]
```diff
@@ -255,6 +278,156 @@ public void testExceptionalDest() throws Exception {
...
+    short containerId = Utils.getRandomShort(TestUtils.RANDOM);
+    boolean toEncrypt = TestUtils.RANDOM.nextBoolean();
+    List<BlobId> blobIdList = new ArrayList<>();
+    for (int i = 0; i < 6; i++) {
+      blobIdList.add(
+          new BlobId(blobIdVersion, BlobId.BlobIdType.NATIVE, ClusterMapUtils.UNKNOWN_DATACENTER_ID, accountId,
+              containerId, partitionId, toEncrypt, BlobId.BlobDataType.DATACHUNK));
+    }
+
+    // Set up VCR
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java]: Can any of this be merged into setupCloudStore or setBasicProperties, or is it too specific to the test?

File: ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java [hunk 6]
```diff
@@ -255,6 +278,156 @@ public void testExceptionalDest() throws Exception {
...
+        new ReplicaThread("threadtest", replicasToReplicate, new MockFindToken.MockFindTokenFactory(), clusterMap,
+            new AtomicInteger(0), cloudDataNode, connectionPool, replicationConfig, replicationMetrics, null,
+            storeKeyConverter, transformer, clusterMap.getMetricRegistry(), false, cloudDataNode.getDatacenterName(),
+            new ResponseHandler(clusterMap), new MockTime());
+
+    long referenceTime = System.currentTimeMillis();
+    // Case 1: Put already expired. Replication happens after Put and after TtlUpdate.
+    // Upload to Cloud only after replicating ttlUpdate.
+    BlobId id = blobIdList.get(0);
+    addPutMessagesToReplicasOfPartition(id, accountId, containerId, partitionId, Collections.singletonList(remoteHost),
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java]: You can move the next 7 lines into a private method and call it for all the cases.

File: ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java [hunk 7]
```diff
@@ -255,6 +278,156 @@ public void testExceptionalDest() throws Exception {
...
+        referenceTime - 2000, referenceTime - 1000);
+    replicaThread.replicate(new ArrayList<>(replicasToReplicate.values()));
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java]: You can call replicasToReplicate.values() once and pass result the the refactored method.

File: ambry-replication/src/test/java/com.github.ambry.replication/InMemoryStore.java [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+  @Override
+  public boolean isEmpty() {
+    return log.blobs.isEmpty();
+  }
+
+  @Override
+  public void shutdown() throws StoreException {
+  }
+}
```
Comment (lightningrob) [inline ambry-replication/src/test/java/com.github.ambry.replication/InMemoryStore.java]: newline at end

File: ambry-replication/src/test/java/com.github.ambry.replication/InMemoryStore.java [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+import java.util.List;
+import java.util.Set;
+
+import static com.github.ambry.replication.ReplicationTest.*;
+
+
+/**
+ * A mock implementation of {@link Store} that store all details in memory.
+ */
+class InMemoryStore implements Store {
```
Comment (lightningrob) [inline ambry-replication/src/test/java/com.github.ambry.replication/InMemoryStore.java]: Should this class be in replication or store package?

File: ambry-replication/src/test/java/com.github.ambry.replication/MockHost.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+
+/**
+ * Representation of a host. Contains all the data for all partitions.
+ */
+public class MockHost {
```
Comment (lightningrob) [inline ambry-replication/src/test/java/com.github.ambry.replication/MockHost.java]: InMemoryHost?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java (hunk 4 above)
Comment (zzmao) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java]: metric added.

File: ambry-replication/src/test/java/com.github.ambry.replication/InMemoryStore.java (hunk 9 above)
Comment (zzmao) [reply ambry-replication/src/test/java/com.github.ambry.replication/InMemoryStore.java]: InMemoryStore is to mock store, classes in store are to test store. So I prefer to leave it here until someone else need it. 
Good point.

File: ambry-replication/src/test/java/com.github.ambry.replication/MockHost.java (hunk 10 above)
Comment (zzmao) [reply ambry-replication/src/test/java/com.github.ambry.replication/MockHost.java]: I struggled here. This is not true in memory, it just a wrapper of `InMemoryStores`.
AND when `InMemory` is used, it usually related to storage, but host here is not.

File: ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java (hunk 6 above)
Comment (zzmao) [reply ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java]: The problem is some case need replicate twice while others just once, and need to assert based on context.

File: ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java (hunk 5 above)
Comment (zzmao) [reply ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java]: Fixed.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java (hunk 3 above)
Comment (zzmao) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java]: added.

File: ambry-replication/src/test/java/com.github.ambry.replication/MockHost.java (hunk 10 above)
Comment (zzmao) [reply ambry-replication/src/test/java/com.github.ambry.replication/MockHost.java]: A similar case is `MockConnectionPool`.

**Pull request merged**